server/grpc_service.go

```diff
@@ -176,6 +183,49 @@ func (s *GrpcServer) Tso(stream pdpb.PD_TsoServer) error {
 		}
 
 		streamCtx := stream.Context()
+		if !s.IsTSOEnabled() {
+			_, forwardedHost, err := s.GetServicePrimaryAddr(ctx, "tso")
+			if err != nil {
+				return errors.WithStack(err)
+			}
+			if forwardStream == nil || lastForwardedHost != forwardedHost {
+				if cancel != nil {
+					cancel()
+				}
+				client, err := s.getDelegateClient(s.ctx, forwardedHost)
+				if err != nil {
+					return err
+				}
+				log.Info("create tso mcs client forward stream", zap.String("forwarded-host", forwardedHost))
+				forwardStream, cancel, err = s.createMCSTSOForwardStream(client)
+				if err != nil {
+					return err
+				}
+				lastForwardedHost = forwardedHost
+				errCh = make(chan error, 1)
+				go forwardMCSTSOClientToServer(forwardStream, server, errCh)
+			}
+			header := request.GetHeader()
+			forwardRequest := &tsopb.TsoRequest{
+				Header: &tsopb.RequestHeader{
+					ClusterId: s.clusterID,
+					SenderId:  header.GetSenderId(),
+					// TODO: whether set keyspace info to 0
+				},
+				Count:      request.GetCount(),
+				DcLocation: request.GetDcLocation(),
+			}
+			if err := forwardStream.Send(forwardRequest); err != nil {
```

Comment:
what will happen if the first forwardStream.Recv() in the goroutine forwardMCSTSOClientToServer()  is executed before the first forwardStream() is called here?

Reply:
I asked this question "regarding grpc.ClientStream, can we call recv() before calling send()?" on ChatGPT. Here is the answer -- No, it is not recommended to call recv() on a grpc.ClientStream before calling send(). In fact, the grpc.ClientStream API is designed such that you must call send() before calling recv().